File: js/src/canon.js
// BoneMesh restricted-JCS certificate canonicalization (security.md §11.1): the
// exact byte string the mesh root signs. Byte-for-byte identical to the Java,
// Elixir, Rust, and Go canonicalizers over the shared corpus
// (spec/corpus/canon.json).
//
// Keys sort by UTF-16 code unit — which is JavaScript's default string ordering,
// so Array.prototype.sort() is already correct. Values may only be strings,
// non-negative integers, or nested objects; anything else is rejected.

export function canonicalize(cert) {
  const filtered = {};
  for (const k of Object.keys(cert)) {
    if (k !== 'sig') filtered[k] = cert[k];
  }
  return encodeObject(filtered);
}

function encodeObject(obj) {
  const keys = Object.keys(obj).sort();
  let out = '{';
  for (let i = 0; i < keys.length; i++) {
    if (i > 0) out += ',';
    out += encodeString(keys[i]);
    out += ':';
    out += encodeValue(obj[keys[i]]);
  }
  return out + '}';
}

function encodeValue(v) {
  if (typeof v === 'string') return encodeString(v);
  if (typeof v === 'number') {
    if (!Number.isInteger(v)) throw new Error(`canon: ${v} is not an integer`);
    if (v < 0) throw new Error(`canon: negative integer ${v}`);
    return String(v);
  }
  if (v !== null && typeof v === 'object' && !Array.isArray(v)) {
    return encodeObject(v);
  }
  throw new Error(`canon: value type ${Array.isArray(v) ? 'array' : typeof v} not permitted in a certificate`);
}

function encodeString(s) {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    switch (c) {
      case 0x22: out += '\\"'; break;
      case 0x5c: out += '\\\\'; break;
      case 0x08: out += '\\b'; break;
      case 0x09: out += '\\t'; break;
      case 0x0a: out += '\\n'; break;
      case 0x0c: out += '\\f'; break;
      case 0x0d: out += '\\r'; break;
      default:
        if (c < 0x20) {
          out += '\\u' + c.toString(16).padStart(4, '0');
        } else {
          out += s[i];
        }
    }
  }
  return out + '"';
}

File: rust/src/canon.rs
//! The BoneMesh restricted-JCS certificate canonicalization (security.md
//! §11.1): the exact byte string the mesh root signs. Byte-for-byte identical
//! to the Java, Go, and Elixir implementations over the shared corpus
//! (spec/corpus/canon.json) — cross-language signature verification depends on
//! it.
//!
//! A certificate contains only JSON strings and non-negative integers, so this
//! is a small subset of RFC 8785: keys sorted by UTF-16 code unit, minimal
//! string escaping, integers in shortest form.

use serde_json::{Map, Value};

/// Canonicalizes a certificate object, dropping any `sig` member. Returns the
/// canonical UTF-8 string, or an error if the object contains a value a
/// certificate may not (float, negative, bool, null, array).
pub fn canonicalize(cert: &Value) -> Result<String, String> {
    let obj = cert.as_object().ok_or("certificate is not an object")?;
    encode_object(obj)
}

fn encode_object(obj: &Map<String, Value>) -> Result<String, String> {
    let mut keys: Vec<&String> = obj.keys().filter(|k| k.as_str() != "sig").collect();
    keys.sort_by(|a, b| utf16(a).cmp(&utf16(b)));

    let mut out = String::from("{");
    for (i, k) in keys.iter().enumerate() {
        if i > 0 {
            out.push(',');
        }
        encode_string(&mut out, k);
        out.push(':');
        encode_value(&mut out, &obj[k.as_str()])?;
    }
    out.push('}');
    Ok(out)
}

fn encode_value(out: &mut String, v: &Value) -> Result<(), String> {
    match v {
        Value::String(s) => {
            encode_string(out, s);
            Ok(())
        }
        Value::Number(n) => match n.as_u64() {
            Some(u) => {
                out.push_str(&u.to_string());
                Ok(())
            }
            None => Err(format!("certificate number {n} is not a non-negative integer")),
        },
        Value::Object(o) => {
            out.push_str(&encode_object(o)?);
            Ok(())
        }
        other => Err(format!("value {other} is not permitted in a certificate")),
    }
}

fn encode_string(out: &mut String, s: &str) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{0008}' => out.push_str("\\b"),
            '\u{0009}' => out.push_str("\\t"),
            '\u{000a}' => out.push_str("\\n"),
            '\u{000c}' => out.push_str("\\f"),
            '\u{000d}' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
}

fn utf16(s: &str) -> Vec<u16> {
    s.encode_utf16().collect()
}
